src/Compilers/CSharp/Portable/Compiler/Instrumentation.cs

```diff
@@ -0,0 +1,212 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.CodeGen;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    internal static class Instrumentation
+    {
+        internal static BoundBlock InjectInstrumentation(MethodSymbol method, BoundBlock methodBody, int methodOrdinal, TypeCompilationState compilationState, CSharpCompilation compilation, DiagnosticBag diagnostics, DebugDocumentProvider debugDocumentProvider, out ImmutableArray<SourceSpan> dynamicAnalysisSpans)
+        {
+            if (methodBody != null)
+            {
+                MethodSymbol createPayload = GetCreatePayload(compilation, diagnostics);
+                MethodSymbol flushPayload = GetFlushPayload(compilation, diagnostics);
+
+                // Do not instrument the instrumentation helpers if they are part of the current compilation (which occurs only during testing). GetCreatePayload will fail with an infinite recursion if it is instrumented.
+                if ((object)createPayload != null && (object)flushPayload != null && (object)method != createPayload && (object)method != flushPayload)
+                {
+                    // Create the symbol for the instrumentation payload.
+                    SyntheticBoundNodeFactory factory = new SyntheticBoundNodeFactory(method, methodBody.Syntax, compilationState, diagnostics);
+                    TypeSymbol boolType = compilation.GetSpecialType(SpecialType.System_Boolean);
```

Comment:
Sometimes we rely on the fact that "goodness" of a type is checked during binding because feature relies on it elsewhere, which is not the case here.